chain/client/src/sync/epoch.rs

```diff
@@ -418,59 +425,160 @@ impl EpochSync {
         epoch_ids.reverse();
 
         // Now that we have all epochs, we can fetch the data needed for each epoch.
-        let epochs = (0..epoch_ids.len() - 2)
+        let epochs = (0..epoch_ids.len() - 1)
             .into_par_iter()
-            .map(|index| -> Result<EpochSyncProofPastEpochData, Error> {
-                let next_next_epoch_id = epoch_ids[index + 2];
+            .map(|index| -> Result<EpochSyncProofEpochData, Error> {
+                let next_epoch_id = epoch_ids[index + 1];
                 let epoch_id = epoch_ids[index];
-                let last_block_header = store
-                    .get_ser::<BlockHeader>(DBCol::BlockHeader, next_next_epoch_id.0.as_bytes())?
-                    .ok_or_else(|| {
-                        Error::Other(format!(
-                            "Could not find last block header for epoch {:?}",
-                            epoch_id
-                        ))
-                    })?;
-                let second_last_block_header = store
-                    .get_ser::<BlockHeader>(
-                        DBCol::BlockHeader,
-                        last_block_header.prev_hash().as_bytes(),
-                    )?
-                    .ok_or_else(|| {
-                        Error::Other(format!(
-                            "Could not find second last block header for epoch {:?}",
-                            epoch_id
-                        ))
-                    })?;
-                let third_last_block_header = store
-                    .get_ser::<BlockHeader>(
-                        DBCol::BlockHeader,
-                        second_last_block_header.prev_hash().as_bytes(),
-                    )?
-                    .ok_or_else(|| {
-                        Error::Other(format!(
-                            "Could not find third last block header for epoch {:?}",
-                            epoch_id
-                        ))
-                    })?;
+                let prev_epoch_id = if index == 0 { after_epoch } else { epoch_ids[index - 1] };
+
+                let (last_final_block_header, approvals_for_last_final_block) =
+                    if index + 2 < epoch_ids.len() {
+                        let next_next_epoch_id = epoch_ids[index + 2];
+                        let last_block_header = store
```

Comment:
nit: separate this logic into a method to reduce nesting.
Maybe it can just accept `store` and `next_next_epoch_id`.

Reply:
I completely agree. But I think this is better done outside of this PR to reduce the amount of changes.

Reply:
@shreyan-gupta this is an example where a store adapter would help a lot. ChainStore and ReadOnlyChunkStore has these helpers but not on Store.